### PR TITLE
TUI: Follow output when scrolled to the bottom

### LINKED ISSUE #209: [DUX-1987] TUI: Follow output when scrolled to bottom
<sub>From [SyncLinear.com](https://synclinear.com) | [DUX-1987](https://linear.app/mercury/issue/DUX-1987/tui-follow-output-when-scrolled-to-bottom)</sub>

### PR DESCRIPTION
**Pull request by 9999years**
Fixes #209.

### CODE REVIEW

File: src/tui/mod.rs [hunk 1]
```diff
@@ -122,10 +122,18 @@ impl Tui {
         self.scroll_offset = min(self.scroll_max(), scroll_offset);
     }
 
+    fn maybe_follow(&mut self) {
+        let height = self.size.height as usize;
+        if self.scroll_offset >= self.line_count - height - 1 {
```
Comment (evanrelf) [inline src/tui/mod.rs:127]: Nit pick: I would add parens to make order of operations more obvious.

```suggestion
        if self.scroll_offset >= self.line_count - (height + 1) {
```

Comment (evanrelf) [inline src/tui/mod.rs:127]: What if `height` is greater than `line_count` (e.g. when first starting)? Might need some saturating arithmetic or some other check here.

File: src/tui/mod.rs [hunk 2]
```diff
@@ -122,10 +122,18 @@ impl Tui {
...
+            self.scroll_offset += 1;
```
Comment (evanrelf) [inline src/tui/mod.rs:128]: I think you're using `saturating_add` in other places, but not here?

**Review by evanrelf — CHANGES_REQUESTED** [review]

File: src/tui/mod.rs [hunk 3]
```diff
@@ -36,17 +36,17 @@ const SCROLL_AMOUNT: usize = 3;
...
 }
 
 impl Default for TuiState {
     fn default() -> Self {
         Self {
             quit: false,
             scrollback: Vec::with_capacity(TUI_SCROLLBACK_CAPACITY),
-            line_count: 1,
-            scroll_offset: 0,
+            line_count: Saturating(1),
```
Comment (evanrelf) [inline src/tui/mod.rs:48]: lol you could go even further and make this a `NonZeroUsize`

Comment (9999years) [reply src/tui/mod.rs:48]: `NonZeroUsize` isn't saturating, weirdly: https://doc.rust-lang.org/stable/core/num/struct.NonZeroUsize.html#method.saturating_add

**Pull request merged**